weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -1390,3 +1464,350 @@ def find_all_descendants(root_ids: typing.List[str]) -> typing.Set[str]:
+        elif isinstance(operation, tsi_query.EqOperation):
+            lhs_part = process_operand(operation.eq_[0])
+            rhs_part = process_operand(operation.eq_[1])
+            cond = f"({lhs_part} = {rhs_part})"
+        elif isinstance(operation, tsi_query.GtOperation):
+            lhs_part = process_operand(operation.gt_[0])
+            rhs_part = process_operand(operation.gt_[1])
+            cond = f"({lhs_part} > {rhs_part})"
+        elif isinstance(operation, tsi_query.GteOperation):
+            lhs_part = process_operand(operation.gte_[0])
+            rhs_part = process_operand(operation.gte_[1])
+            cond = f"({lhs_part} >= {rhs_part})"
+        elif isinstance(operation, tsi_query.SubstrOperation):
+            lhs_part = process_operand(operation.substr_[0])
+            rhs_part = process_operand(operation.substr_[1])
+            cond = f"position({lhs_part}, {rhs_part}) > 0"
+        else:
+            raise ValueError(f"Unknown operation type: {operation}")
+
+        return cond
+
+    def process_operand(operand: tsi_query.Operand) -> str:
+        if isinstance(operand, tsi_query.LiteralOperation):
+            return _param_slot(
+                param_builder.add_param(operand.literal_),  # type: ignore
+                _python_value_to_ch_type(operand.literal_),
+            )
+        elif isinstance(operand, tsi_query.GetFieldOperator):
+            (
+                field,
+                _,
+                fields_used,
+            ) = _transform_external_calls_field_to_internal_calls_field(
+                operand.get_field_, None, param_builder
+            )
+            raw_fields_used.update(fields_used)
+            return field
+        elif isinstance(operand, tsi_query.ConvertOperation):
+            field = process_operand(operand.convert_.input)
+            convert_to = operand.convert_.to
+            if convert_to == "int":
+                method = "toInt64OrNull"
+            elif convert_to == "double":
+                method = "toFloat64OrNull"
+            elif convert_to == "bool":
+                method = "toUInt8OrNull"
+            elif convert_to == "string":
+                method = "toString"
+            else:
+                raise ValueError(f"Unknown cast: {convert_to}")
+            return f"{method}({field})"
+        elif isinstance(
+            operand,
+            (
+                tsi_query.AndOperation,
+                tsi_query.OrOperation,
+                tsi_query.NotOperation,
+                tsi_query.EqOperation,
+                tsi_query.GtOperation,
+                tsi_query.GteOperation,
+                tsi_query.SubstrOperation,
+            ),
+        ):
+            return process_operation(operand)
+        else:
+            raise ValueError(f"Unknown operand type: {operand}")
+
+    filter_cond = process_operation(query.expr_)
+
+    conditions.append(filter_cond)
+
+    return conditions, raw_fields_used
```

Comment:
This file is getting really big - would be nice to split it up.

Reply:
I agree - i actually did refactor this, then reverted: https://github.com/wandb/weave/pull/1656/commits/b6668acaf2f6ed7d0853e573c8007c0d761d0635#diff-121d8320484a693356428be5657e6b61d53f293f5c505f49e67ccc91ac09d5e3 ... i was just touching so many things that i thought it would be better to come back and split this file up after